File: projects/15-webrtc-sfu/src/wire.rs
//! Wire helpers — **fully wired**, not a vertical.
//!
//! Two small things every plane needs before the interesting work starts:
//!
//! 1. [`classify`] — the RFC 7983 first-byte demultiplex. One UDP port carries STUN *and*
//!    RTP *and* RTCP (WebRTC muxes them all), so the very first thing the pump does with a
//!    datagram is decide which it is, by its leading byte(s). This is mechanical, so it's
//!    given to you.
//! 2. Zero-copy RTP header **field accessors** ([`RtpView`]). An SFU forwards RTP *without
//!    decoding it* — it reads a handful of header fields (sequence, timestamp, SSRC, marker,
//!    payload type) and, per subscriber, rewrites some of them **in place**. Parsing the RTP
//!    header from scratch was project 14's V1; here it's plumbing, so these read/patch the
//!    fixed 12-byte header by offset. The learning in *this* project is what the rewriter
//!    (V2) *does* with these, not re-deriving the byte layout.

/// What kind of datagram arrived on the muxed UDP port (RFC 7983 demux).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PacketKind {
    /// A STUN message (ICE connectivity check / response) — first byte in `0x00..=0x03`.
    Stun,
    /// An RTP media packet — first byte in `0x80..=0xBF`, payload-type byte *not* 64..=95.
    Rtp,
    /// An RTCP control packet — first byte in `0x80..=0xBF`, second byte's PT in `192..=223`.
    Rtcp,
    /// Anything else (DTLS, ZRTP, TURN channel, or garbage) — dropped by this SFU.
    Unknown,
}

/// The fixed RTP header size, before any CSRC entries or extensions.
pub const RTP_MIN_HEADER: usize = 12;
/// The RTP version this SFU speaks.
pub const RTP_VERSION: u8 = 2;

/// Classify a datagram by its leading bytes (RFC 7983). Cheap, allocation-free, total.
pub fn classify(buf: &[u8]) -> PacketKind {
    match buf.first() {
        Some(&b) if b <= 3 => PacketKind::Stun,
        Some(&b) if (128..=191).contains(&b) => {
            // RTP vs RTCP share the 0x80 band; the second byte disambiguates: an RTCP
            // packet-type sits in 192..=223, everything else in that band is RTP.
            match buf.get(1) {
                Some(&pt) if (192..=223).contains(&pt) => PacketKind::Rtcp,
                Some(_) => PacketKind::Rtp,
                None => PacketKind::Unknown,
            }
        }
        _ => PacketKind::Unknown,
    }
}

/// A read/patch view over one RTP datagram's fixed header. Borrows the buffer; the
/// accessors are pure offset reads and the mutators patch big-endian fields in place.
///
/// This is the SFU's forwarding lens: it never copies the payload, it just reads the fields
/// the routing/rewriting logic (V2) reasons about and stamps new ones on the way out.
pub struct RtpView<'a> {
    buf: &'a mut [u8],
}

impl<'a> RtpView<'a> {
    /// Wrap a datagram, verifying only that it's long enough to hold a fixed header and is
    /// version 2 — the SFU forwards bytes it doesn't fully parse, but it won't index past
    /// the end of a runt datagram.
    pub fn new(buf: &'a mut [u8]) -> Option<Self> {
        if buf.len() < RTP_MIN_HEADER || (buf[0] >> 6) != RTP_VERSION {
            return None;
        }
        Some(Self { buf })
    }

    /// Marker bit (set on the last packet of a frame — a safe simulcast switch boundary).
    pub fn marker(&self) -> bool {
        self.buf[1] & 0x80 != 0
    }

    /// 7-bit payload type.
    pub fn payload_type(&self) -> u8 {
        self.buf[1] & 0x7f
    }

    /// Per-packet sequence number.
    pub fn sequence(&self) -> u16 {
        u16::from_be_bytes([self.buf[2], self.buf[3]])
    }

    /// Media timestamp (clock-rate ticks).
    pub fn timestamp(&self) -> u32 {
        u32::from_be_bytes([self.buf[4], self.buf[5], self.buf[6], self.buf[7]])
    }

    /// Synchronization source id (identifies the origin stream / simulcast layer).
    pub fn ssrc(&self) -> u32 {
        u32::from_be_bytes([self.buf[8], self.buf[9], self.buf[10], self.buf[11]])
    }

    /// Overwrite the sequence number (the per-subscriber rewriter keeps this contiguous).
    pub fn set_sequence(&mut self, seq: u16) {
        self.buf[2..4].copy_from_slice(&seq.to_be_bytes());
    }

    /// Overwrite the media timestamp (rebased per subscriber on a layer switch).
    pub fn set_timestamp(&mut self, ts: u32) {
        self.buf[4..8].copy_from_slice(&ts.to_be_bytes());
    }

    /// Overwrite the SSRC (the SFU presents one stable SSRC per subscriber, hiding layer
    /// switches behind it).
    pub fn set_ssrc(&mut self, ssrc: u32) {
        self.buf[8..12].copy_from_slice(&ssrc.to_be_bytes());
    }
}

File: projects/15-webrtc-sfu/src/webrtc_sfu/wire.py
"""Wire helpers — **fully wired**, not a vertical.

Two small things the media plane needs before any interesting work starts:

1. `classify` — the RFC 7983 first-byte demultiplex. One UDP port carries STUN
   *and* RTP *and* RTCP (WebRTC bundles them all), so the very first thing the
   pump does with a datagram is decide which it is, from its leading bytes. This
   is mechanical, so it is given to you.
2. `RtpPacket` — read/patch accessors over the fixed 12-byte RTP header. An SFU
   forwards RTP *without decoding it*: it reads a handful of header fields and,
   per subscriber, rewrites some of them. Parsing an RTP header from scratch was
   project 14's V1; here it is plumbing. The learning in *this* project is what
   the rewriter (V2) does with these, not re-deriving the byte layout.

## Why `bytearray`, and what it costs you

Rust's `RtpView<'a>` borrowed `&'a mut [u8]` and the compiler proved nobody else
held it. Python has no such proof and no such borrow: `bytes` is immutable, so
"rewrite the header in place" needs a `bytearray`, and every subscriber needs
its *own* copy because every subscriber gets a different sequence number.

That is not an implementation detail, it is the shape of the boss fight. One
ingress packet at 50 subscribers is 50 `bytearray` allocations of ~1200 bytes
each, and at 1.5 Mbps across three layers that is thousands of allocations a
second feeding the garbage collector. When the Crowded Room's forwarding p99
misses 10 ms, this is the first place to look, and
`docs/15-benchmarks.md` is where the finding goes — with the profile that shows
it, not the intuition that suspected it. (Whether a `memoryview` over one buffer
plus a per-subscriber header patch beats a copy per subscriber is a real
experiment, and a good one.)

## Why the `struct.Struct`s are module-level

`struct.pack(">H", x)` re-parses the format string on every call. `_U16.pack(x)`
against a `Struct` compiled once at import does not. On a path that runs per
packet per subscriber, that difference is measurable — and it is the kind of
CPython-specific cost that does not exist in the Rust this file replaces.
"""

from __future__ import annotations

import struct
from enum import StrEnum

from .errors import BadMagicError, TruncatedError

__all__ = [
    "RTP_MIN_HEADER",
    "RTP_VERSION",
    "Address",
    "PacketKind",
    "RtpPacket",
    "classify",
]

type Address = tuple[str, int]
"""A media-plane transport address: `(host, port)`.

Normalised at the pump boundary. `asyncio` hands IPv6 peers a 4-tuple
(`host, port, flowinfo, scope_id`); the extra two are dropped there so
everything downstream — the nomination table, XOR-MAPPED-ADDRESS, the route
lookup — has exactly one address shape to reason about. Dropping the scope id
does mean link-local IPv6 peers are not addressable, which is a documented
limit, not an oversight: state it in `docs/15-design.md` alongside the rest of
the ICE-lite scope.
"""

RTP_MIN_HEADER = 12
"""The fixed RTP header size, before any CSRC entries or extensions."""

RTP_VERSION = 2
"""The RTP version this SFU speaks."""

_U16 = struct.Struct(">H")
_U32 = struct.Struct(">I")


class PacketKind(StrEnum):
    """What kind of datagram arrived on the muxed UDP port (RFC 7983)."""

    STUN = "stun"
    """First byte in `0x00..=0x03`."""

    RTP = "rtp"
    """First byte in `0x80..=0xBF`, second byte's payload type *not* 192..=223."""

    RTCP = "rtcp"
    """First byte in `0x80..=0xBF`, second byte's packet type in `192..=223`."""

    UNKNOWN = "unknown"
    """DTLS, ZRTP, a TURN channel, or garbage — dropped by this SFU."""


def classify(datagram: bytes) -> PacketKind:
    """Classify a datagram by its leading bytes. Cheap, total, allocation-free.

    Total on *every* input including empty, because this runs before any
    validation on bytes that arrived from anyone at all.
    """
    if not datagram:
        return PacketKind.UNKNOWN
    first = datagram[0]
    if first <= 3:
        return PacketKind.STUN
    if 128 <= first <= 191:
        # RTP and RTCP share the 0x80 band; the second byte disambiguates. An
        # RTCP packet type sits in 192..=223, everything else in that band is RTP.
        if len(datagram) < 2:
            return PacketKind.UNKNOWN
        return PacketKind.RTCP if 192 <= datagram[1] <= 223 else PacketKind.RTP
    return PacketKind.UNKNOWN


class RtpPacket:
    """A read/patch view over one RTP datagram's fixed header.

    Wraps a `bytearray` and patches big-endian fields in place; the payload is
    never touched, copied out, or decoded. This is the SFU's forwarding lens.

    Construction validates, and *raises* rather than returning `None`: a runt or
    wrong-version datagram is a `MediaError` like every other bad datagram, and
    the pump's single `except` around dispatch is where it belongs. That also
    means the fan-out validates once, before the per-subscriber loop, instead of
    re-checking the same bytes for every subscriber.
    """

    __slots__ = ("buf",)

    def __init__(self, buf: bytearray) -> None:
        if len(buf) < RTP_MIN_HEADER:
            raise TruncatedError(need=RTP_MIN_HEADER, got=len(buf))
        if (buf[0] >> 6) != RTP_VERSION:
            raise BadMagicError(f"rtp version {buf[0] >> 6}, want {RTP_VERSION}")
        self.buf = buf

    @property
    def marker(self) -> bool:
        """Marker bit — set on the last packet of a frame, a safe switch boundary."""
        return self.buf[1] & 0x80 != 0

    @property
    def payload_type(self) -> int:
        """7-bit payload type."""
        return self.buf[1] & 0x7F

    @property
    def sequence(self) -> int:
        """Per-packet sequence number."""
        return _U16.unpack_from(self.buf, 2)[0]

    @sequence.setter
    def sequence(self, seq: int) -> None:
        """Overwrite it — the per-subscriber rewriter keeps this contiguous."""
        _U16.pack_into(self.buf, 2, seq & 0xFFFF)

    @property
    def timestamp(self) -> int:
        """Media timestamp, in clock-rate ticks."""
        return _U32.unpack_from(self.buf, 4)[0]

    @timestamp.setter
    def timestamp(self, ts: int) -> None:
        """Overwrite it — rebased per subscriber across a layer switch."""
        _U32.pack_into(self.buf, 4, ts & 0xFFFFFFFF)

    @property
    def ssrc(self) -> int:
        """Synchronization source — identifies the origin stream / simulcast layer."""
        return _U32.unpack_from(self.buf, 8)[0]

    @ssrc.setter
    def ssrc(self, ssrc: int) -> None:
        """Overwrite it — the SFU presents one stable SSRC per subscriber,
        hiding layer switches behind it."""
        _U32.pack_into(self.buf, 8, ssrc & 0xFFFFFFFF)
